python/lsst/resources/http.py

```diff
@@ -1137,6 +1218,116 @@ def walk(
+            # the trusted certificate authorities. fsspect client must use
+            # that context to download the file.
+            ssl_context = ssl.create_default_context()
+            if self._config.ca_bundle is not None:
+                ssl_context.load_verify_locations(capath=self._config.ca_bundle)
+            return aiohttp.ClientSession(connector=aiohttp.TCPConnector(ssl=ssl_context), **kwargs)
+
+        # Retrieve a signed URL for a duration of 1 hour. We assume the URL
+        # is to be used only for downloading, to avoid the risk the caller
+        # modify its contents inadvertently.
+        url = self.generate_presigned_get_url(expiration_time_seconds=3_600)
+
+        # We are returning a a fsspec built-in filesystem implemented by
+        # class 'fsspec.implementations.http.HTTPFileSystem'. The constructor
+        # of this class accepts the argument 'block_size'. The default
+        # value is 'fsspec.utils.DEFAULT_BLOCK_SIZE' which is 5 MB.
+        # That seems to be a reasonable block size for downloading files.
+        return fsspec.core.url_to_fs(url, get_client=get_client)
+
+    def _sign_with_macaroon(self, activity: ActivityCaveat, expiration_time_seconds: int) -> str:
+        # dCache and XrootD servers support delivering of macaroons
+        if self.server is None:
+            raise NotImplementedError(f"server for '{self}' does not support signing URLs")
+
+        # dCache and XrootD servers are known to support delivery of macaroons.
+        if self.server not in ("dcache", "xrootd"):
+            raise NotImplementedError(f"server '{self._server}' does not support signing for {self}")
+
+        match activity:
+            case ActivityCaveat.DOWNLOAD:
+                activity_caveat = "DOWNLOAD,LIST"
+            case ActivityCaveat.UPLOAD:
+                activity_caveat = "UPLOAD,LIST"
+
+        # Request a macaroon for the requested activities and duration duration
+        headers = {"Content-Type": "application/macaroon-request"}
+        body = {
+            "caveats": [
+                f"activity:{activity_caveat}",
+            ],
+            "validity": f"PT{expiration_time_seconds}S",
+        }
+        resp = self._post(data=json.dumps(body), headers=headers)
+
+        # We are expecting the body of to be formatted in JSON.
+        # dCache sets the 'Content-Type' of the response to 'application/json'
+        # but XRootD does not set any 'Content-Type' header 8-[
+        try:
+            response_body = json.loads(resp.text)
+            if "macaroon" in response_body:
+                return f"{self}?authz={response_body['macaroon']}"
+            else:
+                raise ValueError(f"could not retrieve macaroon for URL {self}")
+        except json.JSONDecodeError:
+            raise ValueError(f"could not deserialize response to POST request for URL {self}")
+        except Exception as e:
+            raise e
```

Comment:
Since it just re-raises, the behavior would be the same here without the explicit `except Exception`.

Reply:
✅ Done